File: houseRobber.py
# problem: https://leetcode.com/problems/house-robber/
# Runtime: 32 ms, faster than 69.26% of Python3 online submissions for House Robber.
# Memory Usage: 14.1 MB, less than 79.02% of Python3 online submissions for House Robber.

from typing import List


class Solution:
    def rob(self, nums: List[int]) -> int:
        if len(nums) < 3:
            return max(nums)
        if len(nums) == 3:
            return max(nums[0] + nums[2], nums[1])
        nums[2] += nums[0]
        for i in range(3, len(nums)):
            nums[i] = max(nums[i - 1], nums[i] +
                          nums[i - 2], nums[i] + nums[i-3])
        return nums[-1]

File: houseRobber.cpp
// problem: https://leetcode.com/problems/house-robber/description/
// Runtime
// 5 ms
// Beats
// 22.94%
// Memory
// 7.8 MB
// Beats
// 50.26%

#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
  int rob(vector<int> &nums) {
    if (nums.size() == 1)
      return nums[0];
    for (int i = 2; i < nums.size(); ++i) {
      int first = i > 2 ? nums[i - 3] : 0;
      int second = nums[i - 2];
      nums[i] += max(first, second);
    }
    return max(nums[nums.size() - 1], nums[nums.size() - 2]);
  }
};
